Sources/Core/GTMSessionFetcher.m

```diff
@@ -1004,6 +1022,45 @@ - (NSURLSession *)createSessionWithDelegate:(id<NSURLSessionDelegate>)sessionDel
   return session;
 }
 
+// Asynchronously calculates the User-Agent header from |userAgentProvider|, then
+// sets it in |fetchRequest| and continues the request.
+- (void)updateUserAgentAsynchronouslyForRequest:(NSMutableURLRequest *)fetchRequest
+                              userAgentProvider:(id<GTMUserAgentProvider>)userAgentProvider
+                                       mayDelay:(BOOL)mayDelay
+                                   mayAuthorize:(BOOL)mayAuthorize
+                                    mayDecorate:(BOOL)mayDecorate {
+  GTMSESSION_LOG_DEBUG_VERBOSE(
+      @"GTMSessionFetcher fetching User-Agent from GTMUserAgentProvider %@...", _userAgentProvider);
+  __weak __typeof__(self) weakSelf = self;
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    __strong __typeof__(self) strongSelf = weakSelf;
+    if (!strongSelf) {
+      GTMSESSION_LOG_DEBUG_VERBOSE(@"GTMSessionFetcher deallocated before GTMUserAgentProvider "
+                                   @"fetch dispatched, ignoring.");
+      return;
+    }
+    NSString *userAgent = [userAgentProvider userAgent];
+    GTMSESSION_LOG_DEBUG_VERBOSE(@"Fetched User-Agent string: [%@]", userAgent);
```

Comment:
At this point, would it make sense to assert that the provider now says the value is cached?

Reply:
Done.